File: testdata/jennies/builders/builder_delegation/PythonBuilder/builders/builder_delegation.py
import typing
from ..cog import builder as cogbuilder
from ..models import builder_delegation


class DashboardLink(cogbuilder.Builder[builder_delegation.DashboardLink]):
    _internal: builder_delegation.DashboardLink

    def __init__(self):
        self._internal = builder_delegation.DashboardLink()

    def build(self) -> builder_delegation.DashboardLink:
        """
        Builds the object.
        """
        return self._internal    
    
    def title(self, title: str) -> typing.Self:    
        self._internal.title = title
    
        return self
    
    def url(self, url: str) -> typing.Self:    
        self._internal.url = url
    
        return self
    


class Dashboard(cogbuilder.Builder[builder_delegation.Dashboard]):
    _internal: builder_delegation.Dashboard

    def __init__(self):
        self._internal = builder_delegation.Dashboard()

    def build(self) -> builder_delegation.Dashboard:
        """
        Builds the object.
        """
        return self._internal    
    
    def id(self, id_val: int) -> typing.Self:    
        self._internal.id_val = id_val
    
        return self
    
    def title(self, title: str) -> typing.Self:    
        self._internal.title = title
    
        return self
    
    def links(self, links: list[cogbuilder.Builder[builder_delegation.DashboardLink]]) -> typing.Self:    
        """
        will be expanded to []cog.Builder<DashboardLink>
        """
            
        links_resources = [r1.build() for r1 in links]
        self._internal.links = links_resources
    
        return self
    
    def links_of_links(self, links_of_links: list[list[cogbuilder.Builder[builder_delegation.DashboardLink]]]) -> typing.Self:    
        """
        will be expanded to [][]cog.Builder<DashboardLink>
        """
            
        links_of_links_resources = [[r2.build() for r2 in r1] for r1 in links_of_links]
        self._internal.links_of_links = links_of_links_resources
    
        return self
    
    def single_link(self, single_link: cogbuilder.Builder[builder_delegation.DashboardLink]) -> typing.Self:    
        """
        will be expanded to cog.Builder<DashboardLink>
        """
            
        single_link_resource = single_link.build()
        self._internal.single_link = single_link_resource
    
        return self
    

File: testdata/jennies/builders/builder_delegation/RustBuilder/src/builders/builder_delegation.rs
use crate::cog;
use crate::types::builder_delegation::Dashboard;
use crate::types::builder_delegation::DashboardLink;

#[derive(Debug, Clone)]
pub struct DashboardLinkBuilder {
    internal: DashboardLink,
    errors: Vec<cog::BuildError>,
}

impl DashboardLinkBuilder {
    pub fn new() -> Self {
        Self {
            internal: DashboardLink::default(),
            errors: Vec::new(),
        }
    }
}

impl Default for DashboardLinkBuilder {
    fn default() -> Self {
        Self::new()
    }
}

impl DashboardLinkBuilder {
    pub fn title(mut self, title: String) -> Self {
        self.internal.title = title;

        self
    }
}

impl DashboardLinkBuilder {
    pub fn url(mut self, url: String) -> Self {
        self.internal.url = url;

        self
    }
}

impl cog::Builder<DashboardLink> for DashboardLinkBuilder {
    fn build(&self) -> Result<DashboardLink, Vec<cog::BuildError>> {
        if !self.errors.is_empty() {
            return Err(self.errors.clone());
        }

        Ok(self.internal.clone())
    }
}

#[derive(Debug, Clone)]
pub struct DashboardBuilder {
    internal: Dashboard,
    errors: Vec<cog::BuildError>,
}

impl DashboardBuilder {
    pub fn new() -> Self {
        Self {
            internal: Dashboard::default(),
            errors: Vec::new(),
        }
    }
}

impl Default for DashboardBuilder {
    fn default() -> Self {
        Self::new()
    }
}

impl DashboardBuilder {
    pub fn id(mut self, id: i64) -> Self {
        self.internal.id = id;

        self
    }
}

impl DashboardBuilder {
    pub fn title(mut self, title: String) -> Self {
        self.internal.title = title;

        self
    }
}

/// will be expanded to []cog.Builder<DashboardLink>
impl DashboardBuilder {
    pub fn links(mut self, links: Vec<impl cog::Builder<DashboardLink>>) -> Self {
        let mut built0 = Vec::new();
        for item0 in links {
            let built1 = match item0.build() {
                Ok(val) => val,
                Err(mut err) => {
                    self.errors.append(&mut err);
                    return self;
                }
            };
            built0.push(built1);
        }
        self.internal.links = built0;

        self
    }
}

/// will be expanded to [][]cog.Builder<DashboardLink>
impl DashboardBuilder {
    pub fn links_of_links(
        mut self,
        links_of_links: Vec<Vec<impl cog::Builder<DashboardLink>>>,
    ) -> Self {
        let mut built0 = Vec::new();
        for item0 in links_of_links {
            let mut built1 = Vec::new();
            for item1 in item0 {
                let built2 = match item1.build() {
                    Ok(val) => val,
                    Err(mut err) => {
                        self.errors.append(&mut err);
                        return self;
                    }
                };
                built1.push(built2);
            }
            built0.push(built1);
        }
        self.internal.links_of_links = built0;

        self
    }
}

/// will be expanded to cog.Builder<DashboardLink>
impl DashboardBuilder {
    pub fn single_link(mut self, single_link: impl cog::Builder<DashboardLink>) -> Self {
        let built0 = match single_link.build() {
            Ok(val) => val,
            Err(mut err) => {
                self.errors.append(&mut err);
                return self;
            }
        };
        self.internal.single_link = built0;

        self
    }
}

impl cog::Builder<Dashboard> for DashboardBuilder {
    fn build(&self) -> Result<Dashboard, Vec<cog::BuildError>> {
        if !self.errors.is_empty() {
            return Err(self.errors.clone());
        }

        Ok(self.internal.clone())
    }
}
